juju/juju
# [JUJU-169] fix bundle deployment with 2.9 client and older controllers

Fix 1950237 by using ApplicationsInfo to find existing channels, rather than the status output, which doesn't contain the data in pre-2.9 controllers.  This will only be done for pre 2.9 controllers, as is the case when the defaultCharmSchema is charmstore.  Nicely, ApplicationsInfo is a bulk api call.  However this necessitated requiring applications to have a channel from the bundle. To keep compatibility assume stable if channel not specified in the bundle.

Fix 1931739 by updating the error messages to include the application name causing the failure.  Helps finding the problem in the bundle, where multiple applications exist.

Now that ChangeConfig is internal to juju, ensure there is always a logger.

## QA steps

Please try various combinations of juju clients and controllers with deploying bundles and redeploying them with --dry-run to ensure no changes.  Below are reproducers for the bugs resolved.

```console
$ sudo snap refresh juju --channel 2.8/stable
# using the snap juju:
$ juju bootstrap localhost two-eight
$ juju deploy cs:~ubuntu-mirror-charm-maintainers/ubuntu-mirror
Located charm "cs:~ubuntu-mirror-charm-maintainers/ubuntu-mirror-9".
Deploying charm "cs:~ubuntu-mirror-charm-maintainers/ubuntu-mirror-9".
$ juju deploy cs:~ubuntu-mirror-charm-maintainers/ubuntu-mirror --channel candidate ubuntu-mirror-candidate
Located charm "cs:~ubuntu-mirror-charm-maintainers/ubuntu-mirror-9".
Deploying charm "cs:~ubuntu-mirror-charm-maintainers/ubuntu-mirror-9".
$ juju export-bundle > /tmp/bundle.yaml

# edit the bundle to include the candidate channel for ubuntu-mirror-candidate

# with the 2.9 client:
$ juju deploy /tmp/bundle.yaml --dry-run
No changes to apply.
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1950237
https://bugs.launchpad.net/juju/+bug/1931739
